Add test for the bundle validator

NingLin-P opened the pull request:
This PR adds 2 tests for the bundle validator:
- `duplicated_and_stale_bundle_should_be_rejected` testing duplicated and stale bundle will be rejected when submit to tx pool
- `existing_bundle_can_be_resubmitted_to_new_fork` testing bundle that has already been submitted to a fork can be resubmitted to another fork that does not contains this bundle


### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

NingLin-P committed:
Spawn the maintain-bundles-stored-in-last-k worker and txpool-notifications worker for mock primary node

The maintain-bundles-stored-in-last-k worker is necessary for bundle validator, and
the txpool-notifications worker is used to remove the tx of imported/finalized block
from tx pool which is needed for incoming tests.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add test duplicated_and_stale_bundle_should_be_rejected

This commit also set the config transaction_pool.ban_time to 0 to ensure
submit_bundle tx will not immediately rejected by TemporarilyBanned.

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Add test existing_bundle_can_be_resubmitted_to_new_fork

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu reviewed (commented):
LGTM, but the test failed.

NingLin-P committed:
Fix bundle validator tests

This commit fixs the bundle validator tests mainly by ensuring when a primary block
is produced, the recent_stored_bundles info of the bundle validator is also updated.
There are also some minor refactorings.

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on domains/client/domain-executor/src/tests.rs:
```diff
@@ -522,3 +523,172 @@ async fn pallet_domains_unsigned_extrinsics_should_work() {
+            TransactionSource::External,
+            submit_bundle_tx.clone(),
+        )
+        .await;
```
Looks like a helper method `ferdie.submit_transaction(submit_bundle_tx.clone()).await` can be convenient as well, non-blocker.

liuchengxu commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -126,12 +130,50 @@ impl MockPrimaryNode {
+            "maintain-bundles-stored-in-last-k",
+            None,
+            Box::pin(async move {
+                loop {
+                    tokio::select! {
+                        biased;
+                        maybe_block_imported = imported_blocks_stream.next() => {
+                            match maybe_block_imported {
+                                Some(block) => if block.is_new_best {
+                                    bundle_validator.update_recent_stored_bundles(block.hash);
+                                }
+                                None => break,
+                            }
+                        },
+                        maybe_block_importing = block_importing_stream.next() => {
+                            match maybe_block_importing {
+                                Some((_, mut acknowledgement_sender)) => {
+                                    let _ = acknowledgement_sender.send(()).await;
+                                }
+                                None => break,
+                            }
+                        }
+                    }
+                }
+            }),
```
This is different from the one in production, having a comment to explain the difference is nice.

NingLin-P committed:
Add submit_transaction to MockPrimaryNode

Signed-off-by: linning <linningde25@gmail.com>

NingLin-P committed:
Apply review suggestions

Signed-off-by: linning <linningde25@gmail.com>

liuchengxu commented on test/subspace-test-service/src/mock.rs:
```diff
@@ -142,6 +142,10 @@ impl MockPrimaryNode {
 
+        // The `maintain-bundles-stored-in-last-k` worker here is different from the one in the production code
+        // that it subscribes the `block_importing_notification_stream`, which is intended to ensure the bundle
+        // validator's `recent_stored_bundles` info must be updated when a new primary block is produced, this
```
If I understand correctly, this means the update of `recent_stored_bundles` for the incoming primary block must be complete before producing the next primary block.

NingLin-P replied:
That is right.
